hathor/p2p/peer.py

```diff
@@ -406,7 +234,109 @@ async def validate_entrypoint(self, protocol: 'HathorProtocol') -> bool:
 
         return True
 
-    def validate_certificate(self, protocol: 'HathorProtocol') -> bool:
+    def increment_retry_attempt(self, now: int) -> None:
+        """ Updates timestamp for next retry.
+
+        :param now: current timestamp
+        """
+        self.retry_timestamp = now + self.retry_interval
+        self.retry_attempts += 1
+        self.retry_interval = self.retry_interval * self._settings.PEER_CONNECTION_RETRY_INTERVAL_MULTIPLIER
+        if self.retry_interval > self._settings.PEER_CONNECTION_RETRY_MAX_RETRY_INTERVAL:
+            self.retry_interval = self._settings.PEER_CONNECTION_RETRY_MAX_RETRY_INTERVAL
+
+    def reset_retry_timestamp(self) -> None:
+        """ Resets retry values.
+        """
+        self.retry_interval = 5
+        self.retry_timestamp = 0
+        self.retry_attempts = 0
+        self.flags.discard(PeerFlags.RETRIES_EXCEEDED)
+
+    def can_retry(self, now: int) -> bool:
+        """ Return if can retry to connect to self in `now` timestamp
+            We validate if peer already has RETRIES_EXCEEDED flag, or has reached the maximum allowed attempts
+            If not, we check if the timestamp is already a valid one to retry
+        """
+        if now < self.retry_timestamp:
+            return False
+        return True
+
+
+@dataclass
+class PublicPeer(Peer):
+    public_key: rsa.RSAPublicKey
+
+    @override
+    def to_peer(self) -> Peer:
+        """Base method to convert any subclass to a simple Peer."""
+        return Peer(id=self.id, entrypoints=self.entrypoints[:])
+
+    def to_public_peer(self) -> PublicPeer:
+        """Base method to convert a subclass to a PublicPeer."""
+        return self
+
+    @override
+    def to_json(self, include_private_key: bool = False) -> dict[str, Any]:
+        if include_private_key:
+            raise ValueError('cannot include private key')
+        public_der = self.public_key.public_bytes(
+            encoding=serialization.Encoding.DER,
+            format=serialization.PublicFormat.SubjectPublicKeyInfo,
+        )
+        return {
+            **super().to_json(include_private_key=False),
```

Comment:
Should we do `include_private_key=include_private_key`? I mean, it's the same in the current implementation but future changes might need it.

Reply:
No, otherwise it will always fail because the super class don't allow `include_private_key=True`.

Reply:
Got it. Thanks!